### ISSUE TITLE
SpatialIndex.is_empty is unacceptably slow

### ISSUE THREAD

**Issue by ilaripih** [geopandas/geopandas#1523]
Using GeoPandas v0.8.1

I noticed a serious performance problem when I was checking for the existence of a spatial index with `if GeoDataFrame.sindex`. What I thought should be a simple boolean check turned out to be a really expensive function call that counts all the leaves in the r-tree. Switching to `if GeoDataFrame.sindex.is_empty` resulted in the same thing, counting all leaves.

Is there some almost-zero-cost way to check if the spatial index is non-empty? I'm now comparing `sindex` to `None` but I'm not sure if that's good enough.

https://github.com/geopandas/geopandas/blob/master/geopandas/sindex.py#L66

**Comment by martinfleis** [comment]
Thanks for the report. I guess you are right. Note, that if `sindex` is not built before the check, it triggers its creation, which can take time.

 Just to be precise, the code in question is here, but that does not change the fact that we're calling `leaves()` there.

https://github.com/geopandas/geopandas/blob/301418fc44a1dc1dafd6e0fecabc794c0bc8df58/geopandas/sindex.py#L268-L274

@adriangb any idea if we can avoid this with rtree?

@ilaripih if you can, you can use [pygeos backend](https://geopandas.readthedocs.io/en/latest/install.html#using-the-optional-pygeos-dependency) (including sindex), which should not have this problem as it has `size` attribute. That should be significantly faster.

**Comment by ilaripih** [comment]
Thanks for the very quick response. I did successfully work around the problem so I would only switch to pygeos if it's more efficient than rtree overall.

I think it's okay if the `__len__` method is expensive. But in that case the class should also have a `__bool__` method, and that should be fast.

**Comment by jdmcbr** [comment]
@ilaripih Perhaps I'm missing the point of the question, but are you looking for `_sindex_generated`?

**Comment by adriangb** [comment]
That method dates back quite a while, I'd have to look into rtree itself to see if there's a faster way.

One option I can think of off the top of my head is to assume that all valid geometries end up in the tree and just check `RTreeIndex.geometries`, but I'm not sure if we want to assume that we know what geometries rtree will accept or not.

The above solution would still require building the tree. I don't think it makes sense to be able to check the length without building the tree, so if the expensive part is building the tree, there's no way around that.

@ilaripih could you share your workaround? and any way you can share a gist/google collab/other with a reproducible example that I could test with?

**Comment by adriangb** [comment]
> @ilaripih Perhaps I'm missing the point of the question, but are you looking for `_sindex_generated`?

That wouldn't really work since that gets set even if the index is empty (I think), and it's also a private attribute that might be changed by #1444

**Comment by jdmcbr** [comment]
Yeah, there are two points in the original post that are slightly different, and most of the discussion has been on the second. The first was:

> I noticed a serious performance problem when I was checking for the existence of a spatial index

Which is what `_sindex_generated` does. Then final question is about the sindex being empty:

> Is there some almost-zero-cost way to check if the spatial index is non-empty?

Good point that it is a private method, but I guess I'd say that having a public method for checking whether an sindex has been created or not would probably be good.

**Comment by adriangb** [comment]
> Good point that it is a private method, but I guess I'd say that having a public method for checking whether an sindex has been created or not would probably be good.

I think this is a good point. Checking if the index is empty should not be free (as I stated above), but checking if it has been generated could be valuable. I think this should go into #1444, it should be pretty simple to add a public method that checks the private attribute.

**Comment by ilaripih** [comment]
> @ilaripih could you share your workaround? and any way you can share a gist/google collab/other with a reproducible example that I could test with?

My workaround was just using a separate boolean variable (outside the scope of GeoPandas) to track whether or not I had already successfully created the spatial index. So I guess I'm more interested in the spatial index existing and being valid rather than it being non-empty. Calling `rtree.Index.valid()` seems to work as well, but that's also surprisingly expensive.

So this is not a serious issue now that I'm aware of what the bottleneck was. But I thought I'd report it because I think most people would expect an `is_empty()` method to be of very low cost so I only found the problem when I thoroughly profiled my code with cProfile.

**Comment by adriangb** [comment]
@ilaripih we can _maybe_ provide a method to check if the spatial index has been generated yet, but what we wouldn't really be able to do is add a method to check if the generated spatial index is empty \*

\* we could add a quick check to see if the array of input geometries is empty, but I don't know if this would help your use case or not:
```python3
@property
    def is_empty(self):
        leaves = self.leaves()
        if not self.geometry.size or len(self.leaves()[0][1]) == 0:
            return True
        return len(leaves) > 1
```
This would:

1. Immediately return `False` if no valid input geometries were used to build the tree.
2. Call `leaves()` once instead of twice (not sure what effect that will have).

**Comment by ilaripih** [comment]
@adriangb I'd be fine with caching the is_empty() result or checking if the index has been generated, or both. Or just documenting or adding a warning that the is_empty() method can be slow. I've already fixed my code to run well and it was easy to do once I knew what was going on, so I'm happy with whatever decision you guys make.

**Comment by adriangb** [comment]
For your use case, is the `geometry` attribute of the spatial index empty when `is_empty()==True`?

**Comment by ilaripih** [comment]
> For your use case, is the `geometry` attribute of the spatial index empty when `is_empty()==True`?

In my case the spatial index doesn't ever have a `geometry` attribute:

`AttributeError: 'RTreeIndex' object has no attribute 'geometry'`

**Comment by adriangb** [comment]
Sorry, I meant `geometries` https://github.com/geopandas/geopandas/blob/d3228ddb6814515506921ad5c2e6d6f8bd99c9ae/geopandas/sindex.py#L100
